cpp/src/arrow/ipc/ipc-adapter-test.cc

```diff
@@ -67,42 +103,40 @@ TEST_F(TestWriteRowBatch, IntegerRoundTrip) {
   std::shared_ptr<Schema> schema(new Schema({f0, f1}));
 
   // Example data
+  std::shared_ptr<Array> a0, a1;
+  MemoryPool* pool = default_memory_pool();
+  RETURN_NOT_OK(MakeRandomInt32Array(length, false, pool, &a0));
+  RETURN_NOT_OK(MakeRandomInt32Array(length, true, pool, &a1));
+  out->reset(new RowBatch(schema, length, {a0, a1}));
+  return Status::OK();
+}
 
-  auto data = std::make_shared<PoolBuffer>(pool_);
-  ASSERT_OK(data->Resize(length * sizeof(int32_t)));
-  test::rand_uniform_int(length, 0, 0, std::numeric_limits<int32_t>::max(),
-      reinterpret_cast<int32_t*>(data->mutable_data()));
-
-  auto null_bitmap = std::make_shared<PoolBuffer>(pool_);
-  int null_bytes = util::bytes_for_bits(length);
-  ASSERT_OK(null_bitmap->Resize(null_bytes));
-  test::random_bytes(null_bytes, 0, null_bitmap->mutable_data());
-
-  auto a0 = std::make_shared<Int32Array>(length, data);
-  auto a1 = std::make_shared<Int32Array>(
-      length, data, test::bitmap_popcount(null_bitmap->data(), length), null_bitmap);
-
-  RowBatch batch(schema, length, {a0, a1});
-
-  // TODO(wesm): computing memory requirements for a row batch
-  // 64k is plenty of space
-  InitMemoryMap(1 << 16);
-
-  int64_t header_location;
-  ASSERT_OK(WriteRowBatch(mmap_.get(), &batch, 0, &header_location));
-
-  std::shared_ptr<RowBatchReader> result;
-  ASSERT_OK(RowBatchReader::Open(mmap_.get(), header_location, &result));
+Status MakeListRowBatch(std::shared_ptr<RowBatch>* out) {
+  // Make the schema
+  auto f0 = std::make_shared<Field>("f0", LIST_INT32);
+  auto f1 = std::make_shared<Field>("f1", LIST_LIST_INT32);
+  auto f2 = std::make_shared<Field>("f2", INT32);
+  std::shared_ptr<Schema> schema(new Schema({f0, f1, f2}));
 
-  std::shared_ptr<RowBatch> batch_result;
-  ASSERT_OK(result->GetRowBatch(schema, &batch_result));
-  EXPECT_EQ(batch.num_rows(), batch_result->num_rows());
+  // Example data
 
-  for (int i = 0; i < batch.num_columns(); ++i) {
-    EXPECT_TRUE(batch.column(i)->Equals(batch_result->column(i))) << i
-                                                                  << batch.column_name(i);
-  }
+  MemoryPool* pool = default_memory_pool();
+  const int length = 200;
+  std::shared_ptr<Array> leaf_values, list_array, list_list_array, flat_array;
+  RETURN_NOT_OK(MakeRandomInt32Array(1000, true, pool, &leaf_values));
+  RETURN_NOT_OK(MakeRandomListArray(leaf_values, length, pool, &list_array));
+  RETURN_NOT_OK(MakeRandomListArray(list_array, length, pool, &list_list_array));
+  RETURN_NOT_OK(MakeRandomInt32Array(length, true, pool, &flat_array));
+  out->reset(new RowBatch(schema, length, {list_array, list_list_array, flat_array}));
+  return Status::OK();
 }
 
+INSTANTIATE_TEST_CASE_P(RoundTripTests, TestWriteRowBatch,
+    ::testing::Values(&MakeIntRowBatch, &MakeListRowBatch));
+
+// TODO(emkornfield) More tests
+// Test primitive and lists with zero elements
+// Tests lists and primitives with no nulls
+// String  type
```

Comment:
I think I will add tests to cover the first two and file jira to better support string (and refactor the logical string types to make it convenient for reusing the list serialization logic).  Sound reasonable?